apps/labs/posts/building-scipy-with-flang.md

```diff
@@ -0,0 +1,641 @@
+But it made it extraordinarily difficult to do this in a cross-platform way –
+particularly on Windows, it's not common for users to have a compiler
+installed, much less know how to use it.
+
+It's even worse on the maintainer-side, because they suddenly need to become
+experts in the vagaries of different compilers, linkers and runtime libraries
+necessary on different platforms, and with sometimes wildly varying behaviour.
+
+## Python packaging
+
+This situation also happens to be _the_ fundamental problem in Python packaging.
+There's so much complexity hiding under the surface, that actually dealing with
+it in any kind of sane way is really difficult.
+For example, the C & C++ ecosystems never managed to standardise any sort of
+build & distribution mechanism, and since Python is wrapping around a lot of
+C & C++ code, the problems in Python packaging are a _superset_ of what's
+already a hugely pressing issue in C/C++.
+
+The packaging world in Python has had to continually reinvent itself in order
+to solve these issues.
+One of the biggest problems, which got solved through the "wheel" format,
+was to avoid users having to compile something to install a package.
+
+!['A meme about a difficult choice in Python packaging, with a man unable to decide between two buttons labeled "ABI Hell" and "Users need a compiler"'](/posts/building-scipy-with-flang/ABI_meme.jpg)
+
+This comes with some substantial downsides, especially on the maintainer-side,
+but overall, it is a massive improvement over the situation that existed before.
+Nowadays, most packages are installable through wheels from PyPI.
+Aside from not needing a compiler, this also means that the installation goes
+much faster, there's better metadata, and so on.
+
+The tooling to create these wheels (or to do Python packaging in general)
+underwent many iterations as well.
+At one point in this long [evolution](https://www.youtube.com/watch?v=AQsZsgJ30AE&t=200s),
+`distutils` had emerged as the "one way" to solve things, so much so that it
+even got adopted into the standard library.
+
+However, this was not enough for the Scientific Python community, which had
+bigger problems.
+This [blog](http://technicaldiscovery.blogspot.com/2013/12/why-i-promote-conda.html)
+post by one of the founding contributors of NumPy and SciPy is full of relevant
+anecdotes; here's one quote:
+
+> We in the scientific python community have had difficulty and a rocky history
+> with just waiting for the Python.org community to solve the [packaging]
+> problem. With `distutils` for example, we had to essentially re-write most of
+> it (as `numpy.distutils`) in order to support compilation of extensions that
+> needed Fortran-compiled libraries.
+
+Despite being the product of heroic efforts by many very bright people,
+`numpy.distutils` is generally regarded as a pile of hacks, and has been in
+minimal maintenance mode for years.
+In many ways, this is what led to the creation of conda.
```

Comment:
> but the basic issue is that `distutils`/`setuptools` are build systems, while wheels (and eggs) are package formats

I'm probably taking a few too many short-cuts here, but doing justice to all this _without_ walls of text is really hard. In my defense, I did caveat in the beginning that the recap would be "incomplete and opinionated". ;-)

I can maybe highlight Dustin's talk more, which gives a very nice historical walkthrough (minus conda). In any case, I'll attempt a clarification.

> distribution of those formats on PyPI was what fell short and led to Conda (and this predated wheels).

I wasn't around at that time, but from what I can tell, the timing coincided pretty closely. Continuum Analytics was started in 2012, and the wheel spec got [defined](https://peps.python.org/pep-0427/) around that time too. 

My understanding is that the impetus for conda was by far not only about those package formats and their distribution on PyPI. Again from that blog post:
> In fact, numpy.distutils replaces most of the innards of distutils but is still shackled by the architecture and imperative approach to what should fundamentally be a declarative problem.

Hence why conda was moving things from `setup.py` (imperative; specs resolved at execution time) to recipes (declarative dependencies in advance).

Reply:
Realised/remembered a couple things while working on this:
* I had wanted to setup some background for the removal of `distutils` later
* The whole wheel thing came in because I wanted to also call out some successes and not just appear to have an entire section dedicated to how python packaging sucks.
* The flow of the story further up (+ the meme) made it a no-brainer to talk about wheels, as that's what was necessary to get rid of users needing a compiler.

In any case, I took a stab at rewriting that section and I think it's now better. It goes more to the heart of the issue, it works in some allusions to the imperative/declarative divide that spurred the creation of conda, and it relates much more directly to the embedded image.

Would be happy to hear your thoughts.